llvm/include/llvm/Analysis/TargetTransformInfoImpl.h

```diff
@@ -154,30 +154,22 @@ class TargetTransformInfoImplBase {
   virtual bool isNoopAddrSpaceCast(unsigned, unsigned) const { return false; }
 
   // Assuming that the cast between the two given addrspaces is not a noop,
-  // calculate the width of the given mask value so that it can be applied to
-  // the destination addrspace. In case it cannot be applied since the cast
-  // between the two addrspaces is invalid or the mask value is larger than the
-  // resulting addrspace bit-width, return an empty optional.
-  //
-  // Note that this currently expects the addrspaces to be integral. In case one
-  // of them isn't, an empty optional is returned.
-  virtual std::optional<uint64_t>
-  getAddrSpaceCastMaskWidth(unsigned FromAS, unsigned ToAS, Value *MaskOp,
-                            Instruction *I) const {
+  // calculate the known bits of the resulting ptr in the destination addrspace.
+  // The default implementation returns 0 known bits in case either one of the
+  // addrspaces is not integral or the bit size of the source addrspace is
+  // smaller than the bit size of the destination addrspace.
+  virtual KnownBits computeKnownBitsAddrSpaceCast(unsigned FromAS,
+                                                  unsigned ToAS) const {
     if (DL.isNonIntegralAddressSpace(FromAS) ||
         DL.isNonIntegralAddressSpace(ToAS))
-      return std::nullopt;
-    // All valid 64-bit to 32-bit casts work by chopping off the high
-    // bits. Any masking only clearing the low bits will also apply in the new
-    // address space.
-    if (DL.getPointerSizeInBits(FromAS) != 64 ||
-        DL.getPointerSizeInBits(ToAS) != 32)
-      return std::nullopt;
-    // TODO: Do we need to thread more context in here?
-    KnownBits Known = computeKnownBits(MaskOp, DL, nullptr, I);
-    if (Known.countMinLeadingOnes() < 32)
-      return std::nullopt;
-    return 32;
+      return 0;
```

Comment:
If we had the source value, we could provide the bit-exact result for null pointers (though maybe this is redundant when @shiltian finishes adding the null values to the datalayout)